Generic entity preprocess hooks

gertvdb opened the pull request:
Provide an event for entities by bundle and view mode

pdenooijer commented on hook_event_dispatcher.module:
```diff
@@ -221,6 +225,15 @@ function hook_event_dispatcher_preprocess(&$variables, $hook) {
+
+  // Check whether the item being preprocessed is an entity.
+  if (isset($variables['elements']['#entity_type'])) {
+    /* @var \Drupal\hook_event_dispatcher\Manager\HookEventDispatcherManagerInterface $manager */
```
I think you could add this part to the `PreprocessEventService`.

gertvdb replied:
Do you mean just adding the check to the service like so:

  /**
   * Check whether we are preprocessing an entity.
   */
  public function isEntity(array &$variables) {
    return isset($variables['elements']['#entity_type']);
  }

Or dispatching the events within the service like so: 

 $service->createAndDispatchEntityPreprocessEvents($variables);

pdenooijer commented on src/Event/Preprocess/AbstractEntityPreprocessEvent.php:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @package Drupal\hook_event_dispatcher\Event\Preprocess
+ */
+abstract class AbstractEntityPreprocessEvent extends Event implements EventInterface {
```
Why not extend the `AbstractEventVariables`?

I don't see any reason to make this abstract, you could just combine this class and the concrete implementation (`EntityPreprocessEvent`).

pdenooijer commented on src/Event/Preprocess/EntityBundleViewModePreprocessEvent.php:
```diff
@@ -0,0 +1,31 @@
+   * @return string
+   *   The view mode.
+   */
+  public function getViewMode() {
```
This method can be private.

pdenooijer commented on src/Event/Preprocess/EntityBundlePreprocessEvent.php:
```diff
@@ -0,0 +1,31 @@
+   * @return string
+   *   The bundle.
+   */
+  public function getBundle() {
```
This method should be protected and should use the variables to get the bundle.

pdenooijer commented on src/Event/Preprocess/AbstractEntityPreprocessEvent.php:
```diff
@@ -0,0 +1,53 @@
+   * @return \Drupal\Core\Entity\EntityInterface
+   *   The entity.
+   */
+  public function getEntity() {
```
This method can be removed, the variables are suppose to give you the entity.

pdenooijer commented on src/Event/Preprocess/EntityPreprocessEvent.php:
```diff
@@ -0,0 +1,31 @@
+   * @return string
+   *   The entity type id.
+   */
+  public function getEntityTypeId() {
```
This method can be protected and should use the variables to get the id.

pdenooijer commented on src/Event/Preprocess/Variables/EntityEventVariables.php:
```diff
@@ -0,0 +1,32 @@
+  public function getEntityType() {
+    return $this->variables['elements']['#entity_type'];
+  }
+
```
Add a method to get the bundle & view mode, so those can be used by the event.

pdenooijer reviewed (changes requested):
Please add tests, as they proof that your code does what it suppose to do and keep doing so 😉.

gertvdb commented:
Thanks for the feedback, will try to add tests.

gertvdb commented:
After looking into this issue some more I reworked the code to use the provide suggestions as events. This way it's more generic and can be used for all preprocess functions.

With `hook_theme_suggestions_hook` and `hook_theme_suggestions_hook_alter` you can provide extra theme suggestion for every hook in your project.  So what I do is read out these suggestions in the event service and create an event for each provided suggestion. This way whenever a user creates a new suggestion in it's project it will automatically also create an event for that suggestion. 

I had to rework the factories a bit since they use a static function to get the name and the hook.

Will try to add some tests this week.

pdenooijer commented:
@gertvdb I'll try to have a look as soon as possible, but currently I have no time 😢.

gertvdb commented:
@pdenooijer no problem, take your time. I added some tests for the new stuff I added.

robiningelbrecht commented:
@gertvdb Why would you make your class methods private? This does not allow developers to extend your class and override the function if necessary

gertvdb commented:
@robiningelbrecht what do you mean? I've only added one extra private method 'getSuggestion' in the PR, since it's only for use inside of the class to build the event name based on the suggestion that is passed. The PR provides a way to auto fire events for every suggestion provided via `hook_theme_suggestions_hook` or `hook_theme_suggestions_hook_alter`. So if you provide for example `node__article__teaser` as suggestion in one of these hooks, the event `preprocess_node__article__teaser` will be available to watch via an event listener.

pdenooijer commented on hook_event_dispatcher.module:
```diff
@@ -221,6 +221,13 @@ function hook_event_dispatcher_preprocess(&$variables, $hook) {
+
+  /* @var \Drupal\hook_event_dispatcher\Service\PreprocessSuggestionService $suggestionService */
+  $suggestionService = \Drupal::service('preprocess_suggestion.service');
+  $suggestions = $suggestionService->getHookSuggestions($hook, $variables);
```
I think this is a step to much, this will make the whole preprocess event stuff a lot slower. If your have a page with for instance 20 entities (which is low) this service is called 20 times and sending out a lot of events that are probably useless. I think the previous idea of only sending out entity type and more specific entity type with bundle events is good enough for now. The odd use case where you want to go more specific you can just early return in your listener, instead of sending out power 2 or more events then you previously did.

The amount of events send out on a specific preprocess hook should be set at container build time, but not on every preprocess event call. Further more the idea about this module to replace the hook system, so calling the alter module is kinda counter intuitive.

gertvdb replied:
Yeah, you are propably right I was trying to make it more generic so developers could make their own suggestions and automatically an event was created for that suggestion. But that comes with a cost of not knowing how must events are fired, so is indeed it's better to add fixed events.  I will rework the PR back to firing only events for view_mode and bundle.

pdenooijer replied:
Thanks 👍

pdenooijer commented on src/Event/Preprocess/AbstractPreprocessEvent.php:
```diff
@@ -41,6 +51,23 @@ public static function name() {
+   * @return string
+   *   Event dispatcher type.
+   */
+  public function getDispatcherType() {
```
This breaks with the current preprocess classes and why do the if check in this method. When your sending out suggestions you know you are doing this so a specific method is more appropriate.

gertvdb committed:
Rework to fixed events

gertvdb committed:
Fix typo

gertvdb committed:
Split up Bundle and Entity Events

gertvdb committed:
Fix tests

gertvdb committed:
Add extra test for isEntityHook check

gertvdb committed:
Clean up

gertvdb commented:
I've reworked the PR to only fire events based on EntityType and Bundle. 
**preprocess_entity.ENTITY_TYPE** and **preprocess_entity_bundle.ENTITY_TYPE.BUNDLE** are now fired for every content entity. Making ParagraphPreprocessEvent and NodePreprocessEvent obsolete since you can now use for example `preprocess_entity_bundle.node.article`, `preprocess_entity_bundle.paragraph.banner`, ... . I've also added tests for the code I added. 

Take a look, hope you like it :)

pdenooijer commented:
@gertvdb I'll have a look as soon as possible 👍

pdenooijer commented on hook_event_dispatcher.module:
```diff
@@ -221,6 +221,11 @@ function hook_event_dispatcher_preprocess(&$variables, $hook) {
   $service = \Drupal::service('preprocess_event.service');
   $service->createAndDispatchKnownEvent($hook, $variables);
+
+  if ($service->isEntityHook($variables)) {
```
Why not put all this stuff in the createAndDispatchKnownEvent and rename that to createAndDispatchKnownEvents? I try to keep the module file as clean as possible so the class does all the work 🙂.

gertvdb replied:
Oke will do so. I've got a question about the NodePreprocessEvent and the ParagraphPreprocessEvent. Since they are both content entities, they will now have their previous event and 2 extra events (preprocess_entity.node , preprocess_entity.paragraph). Should I remove these classes so all entities are handled the same, or would you rather keep them to not break backwards compatibility?

pdenooijer replied:
Yeah was thinking about that as well, this is something we should fix with backwards compatibility in mind. Is there not a way to send out the more specific Node & Paragraph entity and just extend them from your new entity event/variables?

gertvdb replied:
I can do something like this :     
`$factory = $this->mapper->getFactory($hook) ? $this->mapper->getFactory($hook) : $this->mapper->getFactory('entity');` and only fire **preprocess_entity.node** when no specific factory is defined. 

So for a node of the type article it will fire **preprocess_node** and **preprocess_entity_bundle.node.article**

And for a unknown entity it will fire **preprocess_entity.MY_ENTITY** and **preprocess_entity_bundle.MY_ENTITY.BUNDLE**

And than make Event/Variables extend the new entity classes. So NodeEventVariables will also have a getEntity, ... method.

I think that would fix backwards compatibility since the event names will stay the same. It will only cause the event name to be inconsistent.

pdenooijer commented on src/Service/PreprocessEventService.php:
```diff
@@ -52,4 +53,38 @@ public function createAndDispatchKnownEvent($hook, array &$variables) {
+   *   A boolean indicating whether an entity hook is called.
+   */
+  public function isEntityHook(array &$variables) {
+    return isset($variables['elements']['#entity_type']) && isset($variables[$variables['elements']['#entity_type']]) && ($variables[$variables['elements']['#entity_type']] instanceof ContentEntityInterface) ? TRUE : FALSE;
```
Please add some new lines after each check so this get more readable.

pdenooijer commented on tests/src/Unit/Preprocess/FactoryMapperTest.php:
```diff
@@ -159,6 +162,66 @@ public function testEckEntityEvent() {
+   * Test a EntityPreprocessEvent.
+   */
+  public function testEntityEvent() {
+    $entityClass = '\Drupal\Core\Entity\ContentEntityInterface';
```
Why not use `ContentEntityInterface::class`?

pdenooijer replied:
Please change all of the occurrences in this PR 🙂.

pdenooijer reviewed (changes requested):
Almost there, just some small changes. Good work 👍

gertvdb committed:
Rework for backwards compatibility

gertvdb commented:
I've reworked some stuff to handle backwards compatibility, the way I suggested above.

gertvdb committed:
Fix test

gertvdb commented:
I've update the PR to handle backwards compatibility the way I suggested above.

gertvdb committed:
Add interface for ComposedPreprocessEvents

gertvdb commented:
@pdenooijer have you gotten a chance to look at changes I've made to handle backwards compatibility?

pdenooijer commented on src/Event/Preprocess/ContentEntityPreprocessEvent.php:
```diff
@@ -0,0 +1,32 @@
+  public function getComposedName() {
+    /** @var \Drupal\hook_event_dispatcher\Event\Preprocess\Variables\ContentEntityEventVariables $vars */
+    $vars = $this->getVariables();
+    return self::name() . '.' . $vars->getEntityType();
```
Won't this make the event `preprocess_node.node`?

gertvdb replied:
This method won't be called for Nodes, Paragraphs, ... due to a check in the service:

`$factory = $factory = $this->mapper->getFactory($hook) ? NULL : $this->mapper->getFactory('entity');`

But thinking about this, it would make more sense to override the getComposedName() on the already existing Content Entity Events and just return `self::name()` there so the check can be removed in the service.

pdenooijer replied:
Yep removing the check would be an improvement 👍, your gonna fix this or should I?

gertvdb replied:
Looked at it some more the reason why I added the check there is that `createAndDispatchKnownEvent()` has already called the node factory and else it would fire 2 events for the node (preprocess_node , preprocess_entity.node).  

But I think I can solve it this way : 

Only fire createAndDispatchKnownEvent() when it's not an entity.
```
/**
   * Create and dispatch the event.
   *
   * @param string $hook
   *   The hook name.
   * @param array $variables
   *   Variables.
   */
  public function createAndDispatchKnownEvents($hook, array &$variables) {
    if ($this->isContentEntityHook($variables)) {
      $this->createAndDispatchEntityEvent($hook, $variables);
      $this->createAndDispatchEntityBundleEvent($variables);
    } 
    else {
      $this->createAndDispatchKnownEvent($hook, $variables);
    }
  }
```

Change the mapper to look first for a more detailed factory and else use the general
```
$factory = $this->mapper->getFactory($hook) ?: $this->mapper->getFactory('entity');
```

Override the `getComposedName();` in all known entities to just return the `self::name();`

pdenooijer commented:
If it's OK with you, I can merge this and rewrite a small part to make sure only one event is send for the entity type when know (node/term/paragraph/etc) or one when unknown (your new generic event). What's your opinion about that 🙂?

gertvdb commented:
Awesome, Perfect for me!

Commit:
improve service

gertvdb commented:
I've added the `getComposedName()` functions to the know entities. Feel free to change some stuff, I think the PR is almost finished :)

Commit:
add test for eck entity

pdenooijer commented:
Seems right to me 👍 , your done committing 😉?

pdenooijer commented:
Thanks for the great work 👍
